cachi2/core/package_managers/rpm/redhat.py

```diff
@@ -0,0 +1,168 @@
+            return False
+        return header.lockfileVendor == "redhat" and header.lockfileVersion == 1
+
+    def process_format(self) -> bool:
+        """Parse the whole lockfile with pydantic and return whether \
+        the parsing was successful and format of the lockfile is correct."""
+        try:
+            self._lockfile = Root(**self.data)
+        except ValidationError as e:
+            log.error(str(e))  # log defail of the validation error
+            return False
+        return bool(self._lockfile)
+
+    def download(self, output_dir: Path) -> None:
+        """
+        Download packages mentioned in the lockfile.
+
+        Go through the parsed lockfile structure and find all RPM and SRPM files.
+        Create a metadata structure indexed by destination path used
+        for later verification (size, checksum) after download.
+        Prepare a list of files to be downloaded, and then download files.
+        """
+        for arch in self._lockfile.arches:
+            log.debug(f"Downloading files for '{arch.arch}' architecture.")
+            # files per URL for downloading packages & sources
+            files: dict[str, Union[str, PathLike[str]]] = {}
+            for pkg in arch.packages:
+                repoid = self.internal_repoid if pkg.repoid is None else pkg.repoid
+                dest = output_dir.joinpath(arch.arch, repoid, Path(pkg.url).name)
+                files[pkg.url] = str(dest)
+                self._files_metadata[dest] = {
+                    "package": True,  # RPM
+                    "url": pkg.url,
+                    "size": pkg.size,
+                    "checksum": pkg.checksum,
+                }
+                Path.mkdir(dest.parent, parents=True, exist_ok=True)
+
+            for pkg in arch.source:
+                repoid = self.internal_source_repoid if pkg.repoid is None else pkg.repoid
+                dest = output_dir.joinpath(arch.arch, repoid, Path(pkg.url).name)
+                files[pkg.url] = str(dest)
+                self._files_metadata[dest] = {
+                    "package": False,  # not package ~ SRPM
+                    "url": pkg.url,
+                    "size": pkg.size,
+                    "checksum": pkg.checksum,
+                }
+                Path.mkdir(dest.parent, parents=True, exist_ok=True)
+
+            asyncio.run(async_download_files(files, get_config().concurrency_limit))
+
+    def verify_downloaded(self) -> None:
+        """Use metadata structure with file sizes and checksums for verification \
+        of downloaded packages and sources."""
+        log.debug("Verification of downloaded files has started.")
+        # check file size and checksum of downloaded files
+        for file_path, file_metadata in self._files_metadata.items():
+            if file_metadata["size"] is not None:  # size is optional
+                if file_path.stat().st_size != file_metadata["size"]:
+                    raise RuntimeError(
```

Comment:
I realized, that `PackageRejected` exception is shown to the user at the end. `RuntimeError` is caught in _main.py_ and converted to _cachi2_ exception. Should I replace such exceptions anyway? I think the current solution saves a few lines because I have a common exception and I just extend the specific message.

Reply:
Yeah, I noticed these exceptions were being treated. I think it would be slightly clearer if you just raised the PackageRejected from here, and that would also allow you to customize each error message individually.